### DISCUSSION TITLE
Assign with promise

### DISCUSSION THREAD

**Discussion by bennyKY** [statelyai/xstate#1785] [Q&A] [answered]
Hi, i have states that need to refresh the context before doing some actions.
For example, i have a state that with delayed event, until the delay has passed the user object might change so at the next state i need to refetch the user (lets assume that i cant listen to the user changed event, not related to the state machine itself obviously).
It looks like this:
```javascript
    signedUp: {
      on: {
        INCOMING_SMS: 'active',
      },
      ...onBoardFlow,
    },
    active: {
      entry: (context, event) => doSomethingWithTheUser(context.user)
      on: {
        PAUSE: 'paused',
      },
    },
```
the delayed event:
```javascript
 send('INCOMING_SMS', { delay: 300000 }); // whole 5 minutes!
```

In the `active` state i have an action `doSomethingWithTheUser` that need to work with the **updated** user - i.e refetch it after the 5 minutes.
Im wondering if adding:
```javascript
entry: [
     assign({ user: (context, event) => await getUser(context.userId)}),  <----
     (context, event) => doSomethingWithTheUser(context.user)
]
```
will solve that? i guess it wont, because the second action will not necessarily run **after** the first `assign` action.
Therefore im using `invoke` service, but it require whole another state which i prefer to avoid because its not a real state rather than an action:
```javascript
    refetchUser: {
      invoke: {
        src: 'getUser',
        onDone: [
          {
            cond: (_context, event) => someCondBasedOnUser(event.data),
            actions: assign({ user: (_context, event) => event.data }),
            target: 'active',
          },
        ],
        onError: { ... }
      },
    },
```

1. How can i achieve it **without** the additional `refetchUser` state?
2. Unrelated - Can i use `cond` inside `invoke` that uses the global `guards` rather than copying it?

**Comment by Andarist** [comment, marked as answer]
> Therefore im using invoke service, but it require whole another state which i prefer to avoid because its not a real state rather than an action:

But it is a state! The state in which the `getUser` is invoked is very distinct from a state when it is already resolved. If we'd support what you are asking for (which I'm pretty sure we won't) there would be no clear way to handle errors. A `getUser` resolving (or erroring) is actually an event and how the machine reacts to that event should be modeled as a transition. You can find some more info in a related discussion here: https://github.com/davidkpiano/xstate/issues/533#issuecomment-510459432

> Unrelated - Can i use cond inside invoke that uses the global guards rather than copying it?

It should "just work". If it isn't then please share a runnable repro case of the problem and I will take a look at it.
